### PR TITLE
`useless_asref`: do not lint `.as_ref().map(Arc::clone)`

### PR DESCRIPTION
**Pull request by samueltardieu**
This applies to `Arc`, `Rc`, and their weak variants. Using `.clone()` would be less idiomatic.

This follows the discussion in <https://github.com/rust-lang/rust-clippy/issues/12528#issuecomment-2014444305>.

changelog: [`useless_asref`]: do not lint `.as_ref().map(Arc::clone)` and similar

### CODE REVIEW

File: clippy_utils/src/ty.rs [hunk 1]
```diff
@@ -159,6 +159,16 @@ pub fn get_type_diagnostic_name(cx: &LateContext<'_>, ty: Ty<'_>) -> Option<Symb
     }
 }
 
+/// Returns true if `ty` is a type on which calling `Clone` through a function instead of
+/// as a method, such as `Arc::clone()`. Lints should avoid suggest replacing instances
```
Comment (y21) [inline clippy_utils/src/ty.rs:163]: Is this sentence missing something at the end? Like "Returns true if `ty` is a type on which [...] **is considered idiomatic**" or something?

**Review by y21 — APPROVED** [review]
Makes sense, good idea to move it to utils

r=me with (or without if it doesn't make sense) doc nit fixed

File: clippy_utils/src/ty.rs (hunk 1 above)
Comment (samueltardieu) [reply clippy_utils/src/ty.rs]: Thanks for catching this.

**Comment by samueltardieu** [comment]
> @bors r=y21

I don't have the needed privileges, maybe you could have used delegate+

**Comment by y21** [comment]
oops, my bad 😅 

Thanks, looks good!

@bors r+

**Pull request merged**